pkg/test/echo/server/forwarder/tcp.go

```diff
@@ -79,27 +81,37 @@ func (c *tcpProtocol) makeRequest(ctx context.Context, req *request) (string, er
 	}
 
 	if _, err := conn.Write([]byte(message + "\n")); err != nil {
-		return outBuffer.String(), err
+		fwLog.Warnf("TCP write failed: %v", err)
+		return msgBuilder.String(), err
 	}
-
+	var resBuffer bytes.Buffer
 	buf := make([]byte, 1024+len(message))
-	n, err := bufio.NewReader(conn).Read(buf)
-	if err != nil {
-		return outBuffer.String(), err
+	for {
+		n, err := conn.Read(buf)
+		if err != nil && err != io.EOF {
+			fwLog.Warnf("TCP read failed (already read %d bytes): %v", len(resBuffer.String()), err)
+			return msgBuilder.String(), err
+		}
+		resBuffer.Write(buf[:n])
+		// the message is sent last - when we get the whole message we can stop reading
+		if err == io.EOF || strings.Contains(resBuffer.String(), message) {
```

Comment:
Will this block forever when response does not contain the sent message? could this happen?

Reply:
A timeout will close the connection giving an EOF or other error.